File: homework/02_homework/file_handling.js
import { parseString } from 'xml2js';
import fs from 'fs';


function removeNestetArrays(obj) {
    for (const key in obj) {
        if (Array.isArray(obj[key]) && obj[key].length === 1) {
            obj[key] = obj[key][0];

            // Convert numeric strings to numbers
            if (!isNaN(obj[key])) {
                obj[key] = Number(obj[key]);
            }
        } else if (typeof obj[key] === 'object') {
            removeNestetArrays(obj[key]);
        }
    }
}


export function parseTxtToJson() {
    try {
        const filePath = '../my_files/my.txt';
        const txtContent = fs.readFileSync(filePath, 'utf8');
        const lines = txtContent.split('\n');

        const result = {};

        lines.forEach(line => {
            // Split each line into key and value
            const [key, value] = line.split(':').map(part => part.trim());

            // Add key-value pair to the result object
            result[key] = value;
        });

        const jsonResult = JSON.stringify(result, null, 2);

        return jsonResult;

    } catch (error) {
        console.error("Error with txt file: ", error);
    }
}


export function parseYamlToJson(){
    try{
        const filePath = '../my_files/my.yaml';
        const yamlContent = fs.readFileSync(filePath, 'utf8');
        const lines = yamlContent.split('\n');

        const jsonResult = {};
        for(const line of lines){
            const [key, value] = line.split(':').map(part => part.trim());

            jsonResult[key] = value;
        }
        
        return JSON.stringify(jsonResult, null, 2);

    }catch(error){
        console.error("Error with yaml file: ", error)
    }
}

export function parseCSVToJson() {
    try {

        const filePath = '../my_files/my.csv';
        const csvContent = fs.readFileSync(filePath, 'utf8');

        const jsonResult = [];

        const lines = csvContent.split('\n');
        const headers = lines[0].split(',');
        
        for (let i = 1; i < lines.length; i++) {
            const currentLine = lines[i].split(',');
            const obj = {};
            for (let j = 0; j < headers.length; j++) {
                obj[headers[j]] = currentLine[j];

                if (headers[j].includes('\r')) {
                    obj[headers[j].replace('\r', '')] = obj[headers[j]];
                    delete obj[headers[j]];
                }
            }
            jsonResult.push(obj);
        }
    
        return JSON.stringify(jsonResult, null, 2);

    } catch (error) {
        console.error("Error with csv file: ", error);
    }
}

export function parseXmlToJson() {
    try {
        const filePath = '../my_files/my.xml';
        const xmlContent = fs.readFileSync(filePath, 'utf8');

        let jsonResult;

        parseString(xmlContent, { trim: true, merge: true }, (err, result) => {
            if (err) {
                throw err;
            }

            // The 'result' is the parsed JavaScript object
            jsonResult = result;

            // Remove arrays from the parsed object
            removeNestetArrays(jsonResult);
        });

        const formattedJsonString = JSON.stringify(jsonResult, null, 2);

        return formattedJsonString;

    } catch (error) {
        console.error("Error with xml file: ", error);
    }
}


export function parseJsonToTxt(){
    try{
        const filePath = '../my_files/my.json';
        const jsonContent = fs.readFileSync(filePath, 'utf8');
        const obj = JSON.parse(jsonContent);
        const result = Object.entries(obj).map(([key, value]) => `${key}: ${value}`).join('\n');
        return result;
    }catch(error){
        console.error("Error with json file: ", error);
    }
}

File: homework/03_homework/file_handling.py
import json
import csv
import yaml
import xml.etree.ElementTree as ET


def parse_txt_to_json():
    try:
        file_path = '../my_files/my.txt'
        with open(file_path, 'r', encoding='utf-8') as file:
            lines = file.readlines()

        result = {}

        for line in lines:
            key, value = map(str.strip, line.split(':'))
            result[key] = value

        json_result = json.dumps(result, indent=2)

        return json_result

    except Exception as error:
        print("Error with txt file:", error)

def parse_yaml_to_json():
    try:
        file_path = '../my_files/my.yaml'
        with open(file_path, 'r', encoding='utf-8') as file:
            yaml_content = file.read()

        json_result = yaml.safe_load(yaml_content)

        return json.dumps(json_result, indent=2)

    except Exception as error:
        print("Error with yaml file:", error)

def parse_csv_to_json():
    try:
        file_path = '../my_files/my.csv'
        with open(file_path, 'r', encoding='utf-8') as file:
            csv_reader = csv.DictReader(file)
            json_result = [row for row in csv_reader]

        return json.dumps(json_result, indent=2)

    except Exception as error:
        print("Error with csv file:", error)

def parse_xml_to_json():
    try:
        file_path = '../my_files/my.xml'
        with open(file_path, 'r', encoding='utf-8') as file:
            xml_content = file.read()

        root = ET.fromstring(xml_content)
        xml_dict = {}

        for child in root:
            xml_dict[child.tag] = child.text

        json_result = json.dumps(xml_dict, indent=2)

        return json_result

    except Exception as error:
        print("Error with xml file:", error)

def parse_json_to_txt():
    try:
        file_path = '../my_files/my.json'
        with open(file_path, 'r', encoding='utf-8') as file:
            json_content = file.read()

        obj = json.loads(json_content)
        result = '\n'.join([f"{key}: {value}" for key, value in obj.items()])

        return result

    except Exception as error:
        print("Error with json file:", error)

def remove_nested_arrays(obj):
    for key, value in obj.items():
        if isinstance(value, list) and len(value) == 1:
            obj[key] = value[0]
            
            # Convert numeric strings to numbers
            if isinstance(obj[key], str) and obj[key].replace('.', '', 1).isdigit():
                obj[key] = float(obj[key])
        elif isinstance(value, dict):
            remove_nested_arrays(value)

